mutagen-core/src/optimistic/add_to_sub.rs

```diff
@@ -0,0 +1,32 @@
+use std::ops::{Add, Sub};
+
+/// trait that is used to optimistically change `+` to `-`.
+///
+/// This trait provides a function `may_sub` that can replace any occurence of `+`.
+/// If the type `Sub` is not implemented for the appropriate types, the function panics.
+pub trait AddToSub<R> {
+    type Output;
+    fn may_sub(self, r: R) -> Self::Output;
+}
+
+impl<L, R> AddToSub<R> for L
+where
+    L: Add<R>,
+{
+    type Output = <L as Add<R>>::Output;
+
+    default fn may_sub(self, _r: R) -> <L as Add<R>>::Output {
+        panic!("not sub");
```

Comment:
Why do we panic here?

Reply:
As far as I understand:

* If the optimistic assumption fails, the mutant should not be considered survived.
* To be able to kill the mutant, the test suite has to fail.
* To fail the test suite, the code must have some different behavior when activating the mutant. 
* I believe panicking is the only/easiest way to cause behavior that is different from the non-mutated case. However, I am open to other options.

In the future, I would like to read the output of the test-binary and detect if a certain panic-message has been printed. Based on that, the mutant could be considered "failed" instead of "killed" or "survived".

Reply:
Ok, that makes sense. I had envisioned a backchannel (an mmap, socket or simple file) to inform the test runner of the ineffective mutation on the no-mutation-run, so the runner could mark the mutation as ineffective. However, there is actually a tri-state at play here, because such a mutation could be within a generic function that monomorphizes to any number of ineffective mutations. Thus, we should be able to both mark an optimistic mutation as (potentially) ineffective and as effective (at least we should do so if it was marked as ineffective before and succeeds nonetheless).

Reply:
I should add that in-band messaging is inherently problematic and I would oppose misusing panics for that purpose. What if a user has a custom panic message that is inadvertently parsed as a killed mutant?

Reply:
I see that using panics might not be the cleanest solution. I will work on some backchannel for coverage-analysis. The information about optimistic assumptions can be included there.

However, I think that not necessarily part of a first working release, which I would like to finish in early September. 

I did not think about optimistic mutantions in generic functions. I will investigate some issues here.

Reply:
I'm fine with that. Currently we just write a file on the nonmutated run, but there are likely faster and leaner solutions.